fdbserver/workloads/MachineAttrition.actor.cpp

```diff
@@ -132,17 +147,117 @@ struct MachineAttritionWorkload : TestWorkload {
 	virtual void getMetrics( vector<PerfMetric>& m ) {
 	}
 
-	struct UIDPredicate {
-		UIDPredicate(StringRef uid ) : uid( uid ) {}
-		bool operator() ( WorkerInterface rhs ) { return rhs.locality.zoneId() != uid; }
-	private:
-		StringRef uid;
-	};
+	static bool noSimIsViableKill(WorkerDetails worker) {
+		if (worker.processClass == ProcessClass::ClassType::TesterClass) return false;
+		return true;
+	}
+
+	ACTOR static Future<Void> noSimMachineKillWorker(MachineAttritionWorkload *self, double meanDelay, Database cx) {
+		ASSERT(!g_network->isSimulated());
+		state int killedMachines = 0;
+		state double delayBeforeKill = deterministicRandom()->random01() * meanDelay;
+		state std::vector<WorkerDetails> allWorkers =
+		    wait(self->dbInfo->get().clusterInterface.getWorkers.getReply(GetWorkersRequest()));
+		// Can reuse reboot request to send to each interface since no reply promise needed
+		state RebootRequest rbReq;
+		if (self->reboot) {
+			rbReq.waitForDuration = self->suspendDuration;
+		} else {
+			rbReq.waitForDuration = std::numeric_limits<uint32_t>::max();
+		}
+		state std::vector<WorkerDetails> workers;
+		// Pre-processing step: remove all testers from list of workers
+		for (const auto& worker : allWorkers) {
+			if (noSimIsViableKill(worker)) {
+				workers.push_back(worker);
+			}
+		}
+		if (self->killDc) {
+			wait(delay(delayBeforeKill));
+			// Pick a dcId to kill
+			deterministicRandom()->randomShuffle(workers);
```

Comment:
This can be moved outside as it is common to all the if conditions.

Reply:
I will move out the shuffle, but I think the delay has to stay because the last if condition runs it in a loop